### PR TITLE
Bash pipelines errors

### PR DESCRIPTION
**Pull request by rainersigwald**
Implement a new bash function in `tools.sh` to emit errors in a Pipelines-compatible way (if in a CI environment). Implementation of #2038 for *nix.

Used them in scripts that had the pattern `echo X >&2`.

Links to show behavior on script failure:

https://github.com/dotnet/arcade/pull/2486/checks?check_run_id=98132257
https://dev.azure.com/dnceng/public/_build/results?buildId=150105&view=logs&jobId=00e6d9d3-6ce7-5b67-7f41-a4c1f1007d9d

This isn't great yet, but I think it's a Pipelines limitation; on the Pipelines Summary and Logs pages the errors are highlighted and formatted well, but they don't translate to the GitHub Checks experience correctly yet--that still just says `Bash exited with code '1'.`

### CODE REVIEW

**Comment by rainersigwald** [comment]
Please don't review yet; trying to tickle CI by moving this out of draft.

**Comment by rainersigwald** [comment]
@jaredpar, @chcosta, I suspect you're interested in this.

**Comment by jaredpar** [comment]
@chcosta what did you just do?

**Comment by chcosta** [comment]
https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops#trigger-builds-using-github-pull-request-comments

File: eng/common/tools.sh [hunk 1]
```diff
@@ -40,6 +42,87 @@ else
...
+  done
+
+  if [[ "$ci" != true ]]; then
+    echo "$@" >&2
+    return
+  fi
+
+  message='##vso[task.logissue'
+
+  if [ -z "$message_type" ]
```
Comment (chcosta) [inline eng/common/tools.sh]: This isn't the normal convention which has "then" on the same line as "if"

File: eng/common/tools.sh [hunk 2]
```diff
@@ -40,6 +42,87 @@ else
...
+  then
+    message_type="error"
```
Comment (chcosta) [inline eng/common/tools.sh]: why not just set "error" as the default above instead of `''`?

File: eng/common/tools.sh [hunk 3]
```diff
@@ -40,6 +42,87 @@ else
...
+        ;;
+      *)
+        break
+        ;;
+    esac
+
+    shift
+  done
+
+  if [[ "$ci" != true ]]; then
```
Comment (chcosta) [inline eng/common/tools.sh]: You could probably move this above the parsing.

File: eng/common/tools.sh [hunk 4]
```diff
@@ -40,6 +42,87 @@ else
...
+  message='##vso[task.logissue'
+
+  if [ -z "$message_type" ]
+  then
+    message_type="error"
+  fi
+
+  message="$message type=$message_type"
+
+  if [ -n "$sourcepath" ]
```
Comment (chcosta) [inline eng/common/tools.sh]: nit: Is anybody going to use these parameters (sourcepath, linenumber, columnnumber, etc...)?  Might just leave them out until they're needed.

Comment (chcosta) [reply eng/common/tools.sh]: I'll reverse on my previous comment, it's probably ok to include this for now and down the road we can re-asses if there's value.

**Comment by chcosta** [comment]
@rainersigwald, this looks pretty good (couple of minor comments).  Any idea when you'll be ready to get this in?

File: eng/common/tools.sh (hunk 4 above)
Comment (rainersigwald) [reply eng/common/tools.sh]: Yeah, I wondered on this. The use case I can think of is if you can extract a better error out of a formatted string emitted by some tool you're running (like, maybe, cmake standard error format?). I can go either way.

Comment (chcosta) [reply eng/common/tools.sh]: I'd say, leave it in.

**Comment by rainersigwald** [comment]
Good feedback. Addressed everything (but left in the extra args).

**Comment by chcosta** [comment]
I thought this was already implemented for Windows, but I don't see it.  @rainersigwald , am I missing something (an active PR somewhere?) or does the Windows implementation of this still need to be done?

**Comment by rainersigwald** [comment]
IIRC there was a sketch of a plan to take the Pipelines logging methods directly for PowerShell, but I don't think anyone has done it.

**Comment by chcosta** [comment]
Thanks Rainer, I'll try to find the issue tracking that and get that addressed.  Any reason not to merge this PR?

**Comment by rainersigwald** [comment]
Nope. Go for it. #2038 covers both so could probably just be left open.

**Pull request merged**